katdal/chunkstore_s3.py

```diff
@@ -62,68 +116,55 @@ class S3ChunkStore(ChunkStore):
-                     client.exceptions.NoSuchKey: ChunkNotFound,
-                     client.exceptions.NoSuchBucket: ChunkNotFound}
+    def __init__(self, session_factory, url):
+        if not requests:
+            raise _requests_import_error
+        try:
+            # Quick smoke test to see if the S3 server is available,
+            # by listing buckets
+            with session_factory() as session:
+                with contextlib.closing(session.get(url)) as response:
+                    _raise_for_status(response)
+        except requests.exceptions.RequestException as error:
+            raise StoreUnavailable(str(error))
+
+        error_map = {requests.exceptions.RequestException: StoreUnavailable,
+                     defusedxml.ElementTree.ParseError: StoreUnavailable}
         super(S3ChunkStore, self).__init__(error_map)
-        self.client = client
+        self._session_pool = _Pool(session_factory)
+        self._url = url
 
     @classmethod
     def _from_url(cls, url, timeout, **kwargs):
         """Construct S3 chunk store from endpoint URL (see :meth:`from_url`)."""
-        if not botocore:
-            raise ImportError('Please install botocore for katdal S3 support')
-        config_kwargs = dict(max_pool_connections=200,
-                             s3={'addressing_style': 'path'})
-        client_kwargs = {}
-        if timeout is not None:
-            config_kwargs['read_timeout'] = int(timeout)
-            config_kwargs['connect_timeout'] = int(timeout)
-            config_kwargs['retries'] = {'max_attempts': 2}
-        # Split keyword arguments into config settings and create_client args
-        for k, v in kwargs.items():
-            if k in botocore.config.Config.OPTION_DEFAULTS:
-                config_kwargs[k] = v
-            else:
-                client_kwargs[k] = v
-        session = botocore.session.get_session()
-        config = botocore.config.Config(**config_kwargs)
-        try:
-            client = session.create_client(service_name='s3',
-                                           endpoint_url=url, config=config,
-                                           **client_kwargs)
-            # Quick smoke test to see if the S3 server is available
-            client.list_buckets()
-        except (ConnectionError, EndpointConnectionError,
-                NoCredentialsError, ClientError, ValueError) as e:
-            raise StoreUnavailable('[{}] {}'.format(type(e).__name__, e))
-        return cls(client)
+        if not requests:
+            raise ImportError('Please install requests for katdal S3 support')
+
+        def session_factory():
+            session = requests.Session()
+            adapter = _TimeoutHTTPAdapter(max_retries=2, timeout=timeout)
+            session.mount(url, adapter)
+            return session
+
+        if kwargs:
+            warnings.warn('Ignoring unknown parameters {}'.format(kwargs.keys()))
```

Comment:
This warning will pop up every time one overrides the capture_block_id and/or stream_name (hopefully not that often), but I don't like warnings when doing something right, especially when the message won't make sense (it knows and applies e.g. `capture_block_id`).

How about registering valid keywords for this kwargs group at the module level (automatically, even)?

Reply:
I just deleted it for now. It would indeed be nice to have some mechanism to warn if someone really does get it wrong.